Generalizing post processing of image files

vinaypotluri opened the pull request:
Image creation script has been modified to make image generation more
generic with support of adding custom finalize scripts in config repo.
This script searches & executes an image's finalize script if it exists.

If any of the image creation steps fail then the stage in pipeline is
set as failed but not the entire pipeline.

Signed-off-by: Vinay Potluri <vinay.potluri@intel.com>

Ref: https://github.com/clearlinux/clr-distro-factory/issues/41

mbelluzzo reviewed (commented):
I had understood that we would provide an entry point for finalize scripts to be **optionally**  called in distro-factory and that the  scripts would live in the config repo where the image files live (that is user defined). That would make more sense IMHO.

So why have you added the finalize scripts and task script on this PR?

vinaypotluri commented:
@mbelluzzo Yes we want them to be part of config repo but by using [post-image] hook in yaml file. The hook isn't supported by clr-installer yet. Until then we can run them from the clr-dsitro-factory repo while generalizing images.sh

Ref: https://github.com/clearlinux/clr-installer/issues/531

mbelluzzo commented:
Config repo does not depend on support of clr-installer. Calling for the scripts in 'images.sh' does not depend on having the .sh on distro-factory. We will use pattern matching to know which scripts needs to be executed.

vinaypotluri committed:
[images] Generalizing post processing of image files

Image creation script has been modified to make image generation more
generic with support of adding custom finalize scripts in config repo.
This script searches & executes an image's finalize script if it exists.

Signed-off-by: Vinay Potluri <vinay.potluri@intel.com>

vinaypotluri committed:
[images] Make successfully creating an image optional

The image stage itself is already optional, i.e. It only gets executed
if there are image definitions in the configuration. But once there are
images to be created, current behavior is to successfully build all the
images in order to consider that stage successful as well.

This patch changes that behavior so one or another image failing to
build is okay since they are considered complimentary artifacts for a
release.

Signed-off-by: Vinay Potluri <vinay.potluri@intel.com>

vinaypotluri commented:
@mbelluzzo As per our discussion, it makes sense to have {image}-post & {image}-finalize scripts in the config repo. So, I will remove the finalize scripts here and modify images.sh to look for {image}-finalize script.

mbelluzzo commented on release/images.sh:
```diff
@@ -66,23 +81,13 @@ create_image() {
+    elif [[ $(echo "${finalize_file}" | wc -w) -eq 1 ]]; then
+        "${finalize_file}"
     fi
+    finalize_images "${name}"
```
it looks like you created a function to finalize all the images but is calling it from a function that create a single image.

vinaypotluri replied:
This was named keeping in mind that some images may have multiple format types. I have renamed to finalize_image to make it less confusing.

mbelluzzo reviewed (changes requested):
see comments

gtkramer commented on release/ReleasePipeline:
```diff
@@ -25,7 +25,9 @@ pipeline {
         stage('Create Images') {
             steps {
-                sh 'release/images.sh'
+                catchError(buildResult: 'SUCCESS', stageResult: 'FAILURE') {
+                    sh 'release/images.sh'
+                }
```
+1, I like this.  It accurately captures how images are an optional part of our build process.

mbelluzzo replied:
This is up to discussion. But being optional or not does not imply that this change is part of generalizing the finalization of scripts, therefore it needs to be moved to its own commit (preferable list the rational for the change as well).

gtkramer commented on release/images.sh:
```diff
@@ -36,6 +36,30 @@ sign_checksum() {
 }
 
+finalize_image() {
+    local name=${1}
```
The interface for this function should provide the full path to the image, because that is how clr-installer will be doing it in the future.  I think this would be an easy change if we do something like:
```
local name_path="${1}"
local name=<some string substitution here to derive the name>
```

Then the rest of the code might be able to stay the same...

vinaypotluri replied:
updated to have name_path

gtkramer commented on release/images.sh:
```diff
@@ -36,6 +36,30 @@ sign_checksum() {
+finalize_image() {
+    local name=${1}
+
+    local finalize_script=$(find "${TEMPLATES_PATH}" -name "${name}-finalize.*" -print -quit)
```
Please quote dollar sign here

vinaypotluri replied:
done

gtkramer commented on release/images.sh:
```diff
@@ -36,6 +36,30 @@ sign_checksum() {
+        if [[ -s "${script_log}_err.log" ]]; then
+            log "Image '${name}'" "NOT OK!"
+            return 1
+        fi
```
You can combine all of this with something like:
```
if ! LOG_METHOD=1 run_and_log "${finalize_script}" "${WORK_DIR}/${name}.img"; then
    log "Finalize step for ${name} image failed" "NOT OK!"
    return 1
fi
```

vinaypotluri replied:
Updated that

gtkramer commented on release/images.sh:
```diff
@@ -36,6 +36,30 @@ sign_checksum() {
+        fi
+    fi
+
+    local images=$(ls "${WORK_DIR}/${name}".*)
```
Please quote dollar sign here

vinaypotluri replied:
done

gtkramer commented on release/images.sh:
```diff
@@ -36,6 +36,30 @@ sign_checksum() {
+        local ext="${image#*.}"
+        mv "${image}" "${final_file}"."${ext}"
+        sha512sum "${final_file}.${ext}" > "${final_file}.${ext}-${CHKSUM_FILE_SUFFIX}"
+        sign_checksum "${final_file}.${ext}-${CHKSUM_FILE_SUFFIX}"
```
`${final_file}.${ext}` is used a few times.  Maybe consider defining a local variable for image file path and reusing it?

vinaypotluri replied:
That has been renamed to use final_path & final_file variables

gtkramer commented on release/images.sh:
```diff
@@ -36,26 +36,49 @@ sign_checksum() {
 
+finalize_image() {
+    local name_path="${1}"
+    local name=$(basename "${name_path}")
```
Please quote your dollar sign here :)

vinaypotluri replied:
done

gtkramer commented:
Aside from this last quoting thing, this lgtm.  +1

mbelluzzo commented on release/images.sh:
```diff
@@ -36,26 +36,49 @@ sign_checksum() {
 
     pushd "${WORK_DIR}" > /dev/null
+
+    set +e
```
prefer to treat the output of the command over doing this.

mbelluzzo replied:
To be honest, this is no-op. All functions  that are executed via parallel are not affect by the top set -e. So you can just drop this change at all.

vinaypotluri replied:
removed

mbelluzzo commented on release/images.sh:
```diff
@@ -65,24 +88,9 @@ create_image() {
         echo
         return 1
     fi
+    set -e
```
ditto

vinaypotluri commented on release/docker-image-creation.sh:
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
```
This file needs to be removed

vinaypotluri commented on release/images.sh:
```diff
@@ -22,40 +22,79 @@ LOG_DIR="${WORK_DIR}/logs"
+    for file in ${file_list}; do
+        local chksum_file="${file}-${CHKSUM_FILE_SUFFIX}"
+
+        sha512sum "${file}" > "${chksum_file}"
```
This generates shasum with full path of file. It can leak unnecessary details. pushd into the directory and then sign the file(s) should prevent that.

mbelluzzo commented:
Merged!
